lib/remap-functions/src/parse_common_log.rs

```diff
@@ -0,0 +1,126 @@
+use remap::prelude::*;
+use std::collections::{BTreeMap, HashMap};
+
+#[derive(Clone, Copy, Debug)]
+pub struct ParseCommonLog;
+
+impl Function for ParseCommonLog {
+    fn identifier(&self) -> &'static str {
+        "parse_common_log"
+    }
+
+    fn parameters(&self) -> &'static [Parameter] {
+        &[Parameter {
+            keyword: "value",
+            accepts: |v| matches!(v, Value::Bytes(_)),
+            required: true,
+        }]
+    }
+
+    fn compile(&self, mut arguments: ArgumentList) -> Result<Box<dyn Expression>> {
+        let value = arguments.required("value")?.boxed();
+
+        Ok(Box::new(ParseCommonLogFn { value }))
+    }
+}
+
+#[derive(Debug, Clone)]
+struct ParseCommonLogFn {
+    value: Box<dyn Expression>,
+}
+
+fn map_fields_to_value(fields: &HashMap<String, LogFieldValue>) -> Value {
+    fields
+        .iter()
+        .map(|(key, value)| {
+            (
+                key.into(),
+                match value {
+                    LogFieldValue::Mapping(mapping) => map_fields_to_value(mapping),
+                    LogFieldValue::Timestamp(timestamp) => Value::Timestamp((*timestamp).into()),
+                    LogFieldValue::Text(text) => Value::Bytes(text.clone().into()),
+                    LogFieldValue::Int(integer) => Value::Integer(*integer as i64),
+                },
+            )
+        })
+        .collect::<BTreeMap<_, _>>()
+        .into()
+}
+
+/// Create a Value::Map from the fields of the given common log event.
+fn log_event_to_value(event: LogEvent) -> Value {
+    map_fields_to_value(event.fields())
+}
+
+impl Expression for ParseCommonLogFn {
+    fn execute(&self, state: &mut state::Program, object: &mut dyn Object) -> Result<Value> {
+        let bytes = self.value.execute(state, object)?.try_bytes()?;
+        let message = String::from_utf8_lossy(&bytes);
+
+        let parser = CommonLogLineParser::new();
+        let parsed = parser
+            .parse(&message)
+            .map_err(|error| format!("Failed parsing common log line: {}", error))?;
+
+        Ok(log_event_to_value(parsed))
+    }
+
+    fn type_def(&self, state: &state::Compiler) -> TypeDef {
+        self.value
+            .type_def(state)
+            .fallible_unless(value::Kind::Bytes)
+            .with_constraint(value::Kind::Map)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::map;
+    use chrono::prelude::*;
+
+    test_function![
+        parse_common_log => ParseCommonLog;
+
+        log_line_valid {
+            args: func_args![value: r#"127.0.0.1 - frank [10/Oct/2000:13:55:36 -0700] "GET /apache_pb.gif HTTP/1.0" 200 2326"#],
+            want: Ok(map![
+                "@timestamp": Value::Timestamp(DateTime::parse_from_rfc3339("2000-10-10T20:55:36Z").unwrap().into()),
+                "@version": "1",
+                "content_length": 2326,
+                "message": "127.0.0.1 - frank [10/Oct/2000:13:55:36 -0700] \"GET /apache_pb.gif HTTP/1.0\" 200 2326",
```

Comment:
Should we retain this? I would think that:

1. We'd either drop this field since it is not represented as structured fields.
2. Use the contained message (`GET /apache_pb.gif HTTP/1.0`)

Reply:
I changed it to include `request_method`, `request_path`, `request_protocol` and `request_line` - in case it does not parse into that format.

Reply:
Maybe I'm unnecessary cautious here and `request_line` will always parse into the `method`, `path`, `protocol` triple? I'm considering to drop the `request_line`.